Signal/src/call/PeerConnectionClient.swift

```diff
@@ -593,177 +678,217 @@ class PeerConnectionClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelD
                 return
             }
-            Logger.debug("\(self.TAG) dataChannel didReceiveMessageWith buffer:\(buffer)")
+            Logger.debug("\(strongSelf.logTag) dataChannel didReceiveMessageWith buffer:\(buffer)")
 
             guard let dataChannelMessage = OWSWebRTCProtosData.parse(from: buffer.data) else {
                 // TODO can't proto parsings throw an exception? Is it just being lost in the Objc->Swift?
-                Logger.error("\(self.TAG) failed to parse dataProto")
+                Logger.error("\(strongSelf.logTag) failed to parse dataProto")
                 return
             }
 
             DispatchQueue.main.async { [weak self] in
-                guard let strongSelf = self else { return }
-                guard let strongDelegate = strongSelf.delegate else { return }
-                strongDelegate.peerConnectionClient(strongSelf, received: dataChannelMessage)
+                completion(dataChannelMessage)
             }
         }
     }
 
     /** The data channel's |bufferedAmount| changed. */
     internal func dataChannel(_ dataChannel: RTCDataChannel, didChangeBufferedAmount amount: UInt64) {
-        Logger.debug("\(TAG) didChangeBufferedAmount: \(amount)")
+        Logger.debug("\(logTag) didChangeBufferedAmount: \(amount)")
     }
 
     // MARK: - RTCPeerConnectionDelegate
 
     /** Called when the SignalingState changed. */
-    internal func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
-        Logger.debug("\(TAG) didChange signalingState:\(stateChanged.debugDescription)")
+    internal func peerConnection(_ peerConnectionParam: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {
+        Logger.debug("\(logTag) didChange signalingState:\(stateChanged.debugDescription)")
     }
 
     /** Called when media is received on a new stream from remote peer. */
-    internal func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
-        guard stream.videoTracks.count > 0 else {
-            return
-        }
-        let remoteVideoTrack = stream.videoTracks[0]
-        Logger.debug("\(self.TAG) didAdd stream:\(stream) video tracks: \(stream.videoTracks.count) audio tracks: \(stream.audioTracks.count)")
+    internal func peerConnection(_ peerConnectionParam: RTCPeerConnection, didAdd stream: RTCMediaStream) {
+        let handle = self.handle
+        let completion: (RTCVideoTrack) -> Void = { [weak self] (remoteVideoTrack) in
+            SwiftAssertIsOnMainThread(#function)
+            guard let strongSelf = handle.get() else { return }
+            guard let strongDelegate = strongSelf.delegate else { return }
 
-        // See the comments on the remoteVideoTrack property.
-        //
-        // We only set the remoteVideoTrack property if peerConnection is non-nil.
-        objc_sync_enter(self)
-        if self.peerConnection != nil {
```

Comment:
This is one of the only places where there are actual changes to the business logic. 

* We don't need to synchronize access to `remoteVideoTrack` property any more.
* We were _unsafely_ accessing `peerConnection` off the signaling queue here - twice.  This has been fixed.  

Overall, this method got a lot simpler and now looks a lot more like the other methods.

Reply:
👍